### ISSUE TITLE
Globbing needs to take care of special file names (versioning, conflicts)

### ISSUE THREAD

**Issue by calmh** [syncthing/syncthing#2606]
Wherever we build glob patterns based on existing file names we need to take care of the fact that they can contain things like `[-]` which gets interpreted as a broken character range.

Probably we need a well tested utility function to build these patterns, or use something other than globbing.

**Comment by nrm21** [comment]
How about just not using dash anymore but maybe underscore instead like ( .*sync_conflict.* )  Then search for that instead?  Or are there other patterns that need to be found with dashes also?

**Comment by kluppy** [comment]
The dash is not the only problem, the file names might contain other characters that have special meaning when pattern matching. The use of square brackets is what actually highlight this problem on the forum.

**Comment by kluppy** [comment]
The Go `regexp` package actually has a function to look after this.

`func QuoteMeta(s string) string`
https://godoc.org/regexp#QuoteMeta

**Comment by calmh** [comment]
Need to check whether that does the right though, as we're not quoting regexps but glob patterns, and the backslash is the path separator... But yeah, something like that.

**Comment by kluppy** [comment]
Your right. The source from Go shows the escape characters are ignored on windows systems.
This seems like laziness in terms of go development to me. What is the point of having pattern matching if they are just going to turn off the escape characters on one of the platforms.
It is now useless to us.

**Comment by AudriusButkevicius** [comment]
Is this still an issue?

**Comment by calmh** [comment]
Yep. https://forum.syncthing.net/t/globbing-syntax-error-in-pattern/8463

**Comment by AudriusButkevicius** [comment]
Glob package we are using has QuoteMeta specifically for this purpose

**Comment by calmh** [comment]
I'm suspicious against how well that works. It adds backslashes in front of meta characters. But backslash is the path separator on Windows. And I don't think it supports escaping on Windows. So it remains broken.

**Comment by AudriusButkevicius** [comment]
We it should be able to handle its own escaping?
Can you give an example where it does not work?

**Comment by calmh** [comment]
You have a file `Movies\2001 - A Space Odyssey [1968]\movie.mp4`. We will add our glob pattern `[0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9]-[0-9][0-9][0-9][0-9][0-9][0-9]` to make `Movies\2001 - A Space Odyssey [1968]\movie~[0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9]-[0-9][0-9][0-9][0-9][0-9][0-9].mp4`

If we do escaping before adding the glob pattern we would get `Movies\2001 - A Space Odyssey \[1968\]\movie~[0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9]-[0-9][0-9][0-9][0-9][0-9][0-9].mp4` which is just two more directory separators in the middle of the name - and for that reason I don't think the glob library uses `\` for escapes on Windows.

We also do a whole bunch of `filepath.Dir()` and `filepath.Join` on the paths both before and after adding the glob pattern which will probably screw up with the scaping added.

**Comment by AudriusButkevicius** [comment]
So I'll give you the filepath.Dir() etc case, yet that should all happen before matching not after, and QuoteMeta actually gives: `Movies\\2001 - A Space Odyssey \[1968\]\\movie.mp4` which I think would work correctly when globbing the final path, which is essentially what we care about?

**Comment by calmh** [comment]
I'm still 99% sure that the backslash escape is only used on Unix, not Windows. It might still solve the problem on Unix though.

The filepath stuff is a little bit convoluted as we split before adding glob patterns and join afterwards, but on Unix it might work since there is no issue with the backslash there.

More generally we should probably just listdir and look through the files for matches ourselves instead of trying to build a bulletproof glob pattern.

**Comment by AudriusButkevicius** [comment]
I produced this output on Windows.

**Comment by calmh** [comment]
I don't doubt that `QuoteMeta` runs on Windows. I doubt that the result does what is intended.

**Comment by maxexcloo** [comment]
Have been having this issue on some files also:

```
 2024-04-08 16:00:12: globbing: syntax error in pattern for Data/Scripts/retool/app/tests/goldens/features/Retool - Exclusions (-x) [-aAbBcdDefkmMopPruv]~[0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9]-[0-9][0-9][0-9][0-9][0-9][0-9].dat

2024-04-08 16:02:12: globbing: syntax error in pattern for Data/Cache/DATs/No-Intro/Bally - Astrocade (Tapes) (Parent-Clone) (20220914-145554) (Retool 2024-04-08 10-38-45) (1) [-AbBcdefkmopPruv]~[0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9]-[0-9][0-9][0-9][0-9][0-9][0-9].dat

2024-04-08 16:02:12: globbing: syntax error in pattern for Data/Cache/DATs/No-Intro/Commodore - Commodore 64 (Parent-Clone) (20240127-100749) (Retool 2024-04-08 10-36-39) (290) [-AbBcdefkmopPruv]~[0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9]-[0-9][0-9][0-9][0-9][0-9][0-9].dat

2024-04-08 16:02:12: globbing: syntax error in pattern for Data/Cache/DATs/No-Intro/VTech - CreatiVision (Parent-Clone) (20230426-080718) (Retool 2024-04-08 10-39-59) (24) [-AbBcdefkmopPruv]~[0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9]-[0-9][0-9][0-9][0-9][0-9][0-9].dat 
```

**Comment by niwazukihon** [comment]
Encountered this issue today. I'm surprised that syncthing has such a huge restriction. There should be a warning somewhere saying that syncthing doesn't work with filenames with characters like `[` and `]`.

**Comment by serge-vk** [comment]
This is still a problem today, as I've just found out. The interesting thing, however, is that I am unable to find the file with the name that generated the error message on any sync'ed computer (a mix of Ubuntu and Mac machines). Probably, it is in the version history, but I am not sure how to check this.

**Comment by JanKanis** [comment]
+1 still an issue
